be/src/storage/storage_engine.cpp

```diff
@@ -1098,6 +1113,92 @@ bool StorageEngine::check_rowset_id_in_unused_rowsets(const RowsetId& rowset_id)
     return search != _unused_rowsets.end();
 }
 
+void StorageEngine::remove_increment_map_by_table_id(int64_t table_id) {
+    {
+        // LIKELY, lock shared mode.
+        std::shared_lock l(_auto_increment_mutex);
+        auto iter = _tableid_auto_increment_map.find(table_id);
+        if (iter == _tableid_auto_increment_map.end()) {
+            return;
+        }
+    }
+
+    // UNLIKELY, lock mode.
+    std::unique_lock l(_auto_increment_mutex);
+    if (_tableid_auto_increment_map.find(table_id) == _tableid_auto_increment_map.end()) {
+        return;
+    }
+    delete _tableid_auto_increment_map[table_id];
+    _tableid_auto_increment_map.erase(_tableid_auto_increment_map.find(table_id));
+}
+
+void StorageEngine::get_next_increment_id_interval(int64_t tableid, size_t num_row,
+                                                   std::vector<int64_t> &ids) {
+    bool need_init = false;
+    // lock shared mode to check <tableid, <min,max>> exist or not.
+    _auto_increment_mutex.lock_shared();
+    auto iter = _tableid_auto_increment_map.find(tableid);
+    auto end_iter = _tableid_auto_increment_map.end();
+    std::pair<int64_t, int64_t> *interval;
+
+    if (UNLIKELY(iter == end_iter)) {
+        _auto_increment_mutex.unlock_shared();
+        // lock mode to add the new <tableid, <min,max>>
+        _auto_increment_mutex.lock();
+
+        if (_tableid_auto_increment_map.find(tableid) == _tableid_auto_increment_map.end()) {
+            _tableid_auto_increment_map.insert({tableid, new std::pair<int64_t,int64_t>({0, 0})});
+            need_init = true;
+        }
+
+        interval = _tableid_auto_increment_map.at(tableid);
+        _auto_increment_mutex.unlock();
+    } else {
+        interval = _tableid_auto_increment_map.at(tableid);
+        _auto_increment_mutex.unlock_shared();
+    }
+
+    // lock shard for different tableid
+    std::unique_lock<std::mutex> l((*_tables_auto_increment_mutex_shards)
+                                   [tableid % _tables_auto_increment_shards_mask]);
+
+    // avaliable id interval: [cur_avaliable_min_id, cur_max_id)
+    int64_t &cur_avaliable_min_id = interval->first;
+    int64_t &cur_max_id = interval->second;
+    CHECK_GE(cur_max_id, cur_avaliable_min_id);
+
+    size_t cur_avaliable_rows = cur_max_id - cur_avaliable_min_id;
+    auto ids_iter = ids.begin();
+    if (UNLIKELY(need_init || num_row > cur_avaliable_rows)) {
+        size_t alloc_rows = num_row - cur_avaliable_rows;
+        // rpc request for the new available interval from fe
+        TAllocateAutoIncrementIdParam alloc_params;
+        TAllocateAutoIncrementIdResult alloc_result;
+
+        alloc_params.__set_table_id(tableid);
+        alloc_params.__set_rows(alloc_rows);
```

Comment:
looks like there aren't any caching here, just alloc the amount needed, is it causing more rpcs?

Reply:
now use a parameter in fe to control the extra id cache for each allocation.